src/erlcloud_aws.erl

```diff
@@ -574,6 +590,51 @@ prop_to_list_defined( Name, Props ) ->
     end.
 
 
+-spec get_role_credentials(aws_config()) -> {ok, #role_credentials{}} | {error, term()}.
+get_role_credentials(#aws_config{assume_role = AssumeRoleOptions} = Config) ->
+    RoleArn = proplists:get_value(role_arn, AssumeRoleOptions, undefined),
+    case application:get_env(erlcloud, {role_credentials, RoleArn}) of
+        {ok, #role_credentials{expiration_gregorian_seconds = Expiration} = Credentials} ->
+            Now = calendar:datetime_to_gregorian_seconds(calendar:universal_time()),
+            %% Get new credentials if these will expire in less than 2 minutes
+            case Expiration - Now < 120 of
+                true -> get_credentials_from_role(Config);
+                false -> {ok, Credentials}
+            end;
+        undefined ->
+            get_credentials_from_role(Config)
+    end.
+
+-spec get_credentials_from_role(aws_config()) -> {ok, #role_credentials{}} |
+                                                 {error, term()}.
+get_credentials_from_role(#aws_config{assume_role = AssumeRoleOptions} = Config) ->
+    RoleArn = proplists:get_value(role_arn, AssumeRoleOptions, undefined),
+    SessionName = proplists:get_value(session_name, AssumeRoleOptions,
+                                      "erlcloud"),
+    SessionExpiration = proplists:get_value(session_expiration,
+                                            AssumeRoleOptions, 900),
+    ExternalId = proplists:get_value(external_id, AssumeRoleOptions,
+                                      undefined),
+    %% We have to remove the assume role options to make sure we do not
+    %% enter in a infinite loop because erlcloud_sts:assume_role also calls
+    %% update_config deep inside when makes the request
+    case catch erlcloud_sts:assume_role(Config#aws_config{assume_role = []},
+                                        RoleArn, SessionName,
+                                        SessionExpiration, ExternalId) of
+        {#aws_config{}=_NewConfig, Creds} ->
+            ExpireAt = calendar:datetime_to_gregorian_seconds(
+                proplists:get_value(expiration, Creds)),
+            Record = #role_credentials{
+                access_key_id =  proplists:get_value(access_key_id, Creds),
+                secret_access_key = proplists:get_value(secret_access_key, Creds),
+                session_token = proplists:get_value(session_token, Creds),
+                expiration_gregorian_seconds = ExpireAt},
+            application:set_env(erlcloud, {role_credentials, RoleArn}, Record),
+            {ok, Record};
+        {'EXIT', Reason} ->
```

Comment:
previous comment not addressed, let's protect for calling the fun with `undefined` and not do `catch`